Ivan Ornes - Feed API Challenge

ivanornes opened the pull request.

mapostolakis commented on FeedAPIChallenge/Feed API/RemoteImagesLoader.swift:
```diff
@@ -0,0 +1,17 @@
+
+import Foundation
+
+internal struct RemoteImagesLoader {
+	
+	internal static func getImages(_ data: Data) throws -> [FeedImage] {
```
Consider the name you chose for this struct. "Loading" usually signifies operations with side-effects such as loading from the network or a database. But `getImages` is a mapping function with no side-effects. 

It maps data to `[FeedImage]` so it would be better to first rename the struct accordingly avoiding the suffix "Loader", i.e. `FeedImageMapper`.

Then you can also rename the function `getImages` to simply `map` to denote its purpose.

mapostolakis replied:
The `throws` you have marked this function with is never utilized.

mapostolakis commented on FeedAPIChallenge/Feed API/RemoteImagesLoader.swift:
```diff
@@ -0,0 +1,17 @@
+internal struct RemoteImagesLoader {
+	
+	internal static func getImages(_ data: Data) throws -> [FeedImage] {
+		guard let dictionary = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [String: Any] else { return [] }
```
Check the return statement of this guard clause.

You're returning an empty array if the data serialization fails. However, an empty `FeedImage` array is a valid feed representation.

It would be best to throw an error communicating to the clients of this function what went wrong, which in this case is "invalid data".

mapostolakis commented on FeedAPIChallenge/Feed API/RemoteImagesLoader.swift:
```diff
@@ -0,0 +1,17 @@
+	
+	internal static func getImages(_ data: Data) throws -> [FeedImage] {
+		guard let dictionary = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [String: Any] else { return [] }
+		return (dictionary["items"] as? [[String: Any]] ?? []).compactMap(RemoteImage.init).map { $0.feedImage }
```
`dictionary["items"] as? [[String: Any]]` is part of the data validation/extraction process. 

If you can't access the items as `[[String: Any]]` then the data should be considered invalid and you should throw an error.

Include this check in the guard statement.

mapostolakis replied:
By using `compactMap` and discarding `nil` results you end up with the same problem you got by returning an empty array in the `guard` statement or the `dictionary["items"] as? [[String: Any]] ?? []`. Check my comments above for a more detailed explanation.

If the data for a JSON feed image is invalid you shouldn't ignore it. Instead, you should notify the clients of this function that the mapping process failed by throwing an appropriate error.

mapostolakis commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -4,21 +4,39 @@
 	public func load(completion: @escaping (FeedLoader.Result) -> Void) {
-		
+		client?.get(from: url) { [weak self] result in
+			guard self?.client != nil else { return }
```
Check if `self` is `nil` instead of the client. I suggest you also review the `test_load_doesNotDeliverResultAfterSUTInstanceHasBeenDeallocated` test in the lectures in case this isn't clear yet.

mapostolakis commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -4,21 +4,39 @@
-		
+		client?.get(from: url) { [weak self] result in
+			guard self?.client != nil else { return }
+			do {
+				switch result {
+				case .success((let data, let response)):
+					if response.statusCode == 200 {
+						let images = try RemoteImagesLoader.getImages(data)
+						completion(.success(images))
+					} else {
+						completion(.failure(Error.invalidData))
+					}
+				case .failure(_): completion(.failure(Error.connectivity))
+				}
+			} catch {
+				completion(.failure(Error.invalidData))
+			}
```
Since it's the `RemoteImagesLoader.getImages` that throws you don't have to wrap the whole `load` body in a `do/catch` block. 

You can wrap only the part of the throwing function, reducing the indentation and making the code more manageable and easier to read.

For example:

```
switch result {
case .success((let data, let response)):
  if response.statusCode == 200 {
    do {
      let images = try RemoteImagesLoader.getImages(data)
      completion(.success(images))
    } catch {
      completion(.failure(Error.invalidData))
    }
  } else {
    completion(.failure(Error.invalidData))
  }
case .failure(_):
  completion(.failure(Error.connectivity))
}

```

mapostolakis reviewed (commented):
Hi @ivanornes,

You almost passed the challenge!

Please, check my review comments and let me know when you're done so I can review it again.

Thank you!

ivanornes commented:
Hi Mike, 

I have just reviewed the code. 

Thanks for the great feedback!

mapostolakis commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -6,19 +6,35 @@ import Foundation
 public final class RemoteFeedLoader: FeedLoader {
 	private let url: URL
-	private let client: HTTPClient
+	private var client: HTTPClient?
```
The goal of the `RemoteFeedLoader` is to request data over HTTP and map it to a native feed. So this component can't function properly without an `HTTPClient`, thus it shouldn't be optional. 

You can revert to `let client: HTTPClient`.

mapostolakis commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -6,19 +6,35 @@ import Foundation
 	}
 	
-	public init(url: URL, client: HTTPClient) {
+	public init(url: URL, client: HTTPClient?) {
```
Please see my comment above and remove the optional.

mapostolakis reviewed (commented):
Hi @ivanornes, almost there!

Please, check my last couple of review comments and let me know once you're done so I can review it again. Thank you!

mapostolakis commented on FeedAPIChallenge/Feed API/FeedImageMapper.swift:
```diff
@@ -0,0 +1,18 @@
+internal struct FeedImageMapper {
+	
+	internal static func map(_ data: Data) throws -> [FeedImage] {
+		guard let dictionary = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [String: Any],
+			  let imageItems = dictionary["items"] as? [[String: Any]] else { throw RemoteFeedLoader.Error.invalidData }
+		return try imageItems.map(RemoteImage.init).map { $0.feedImage }
+	}
```
I suggest you also learn how to use `Codable` as a replacement to `JSONSerialization`. It will help you simplify your code and avoid multiple checks. You can centralize the mapping logic instead of spreading it in multiple places: e.g. this function and `RemoteImage.init`.

ivanornes commented:
Hi Mike, 

I have just made the two changes. I removed the HTTPClient optional and changed the JSONSerialization for JSONDecoder. 

Using the Codable Protocol was my first choice but it didn't worked the first time, so I used the JSONSerialization. Now I made it work with the Decodable protocol. 

Anyway, the code now looks much cleaner. 

Thanks for the feedback!

mapostolakis commented:
Hi @ivanornes,

Congratulations, you've passed the challenge 🥇

All tests are passing, so your implementation complies with the specs.

Great job! ✅

mapostolakis commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -6,7 +6,7 @@ import Foundation
 public final class RemoteFeedLoader: FeedLoader {
 	private let url: URL
-	private let client: HTTPClient
+	private var client: HTTPClient
```
Revert to: 

```swift
private let client: HTTPClient
```

ivanornes replied:
Hi @mapostolakis,

Reverted the var to let. 

Thanks Mike! 😊
